persist IBD latch across restarts

pstratem opened the pull request:
In the unlikely event that all or most of the listening nodes are restarted
into Initial Block Download mode the network might get stuck with little or
no listening nodes providing headers or blocks.

The timeout is intended to move nodes out of IBD so they will request and
serve blocks and headers to inbound peers.

Edit: By latching IBD across restarts we prevent the network from getting into this state at all.

jnewbery commented on src/validation.cpp:
```diff
@@ -1305,6 +1305,14 @@ bool CChainState::IsInitialBlockDownload() const
     if (fImporting || fReindex)
         return true;
+    uint64_t ibdtimeout = gArgs.GetArg("-ibdtimeout", DEFAULT_IBD_TIMEOUT);
+    int64_t uptime = GetTime() - GetStartupTime();
+    assert(uptime >= 0);
```
This assumes that `GetTime()` is monotonic, which I don't think is true. `GetTime()` is mockable so may move backwards in tests.

It looks like `GetTime()` is also deprecated.

pstratem replied:
Nodes in IBD will not respond to GETHEADERS. Nodes in IBD do not request headers from inbound peers only outbound.

If all or most of the listening nodes are in IBD (emergency hot fix or something) then nobody will be able to get headers.

To get out of this the listening nodes need to request headers from inbound connections, either by exiting IBD or special casing that limit while in IBD.

schuie2528 replied:
bc1que90l5dy4yzn79x8xmem8t7dgn0c2zcqvxnr6x

jnewbery reviewed (commented):
I think this is essentially saying that if the node can't get out of IBD within two days, then just pretend we've completed IBD anyway. I don't think that's the right behaviour - if we've stalled and aren't making any progress towards getting out of IBD, we'd be better off warning the user very loudly and then perhaps shutting down.

MarcoFalke commented:
> all or most of the listening nodes are restarted into Initial Block Download mode 

Wouldn't that mean that there hasn't been a block in 24 hours? Then, waiting another two days for the next block doesn't seem like a fallback that will be needed.

pstratem commented:
@MarcoFalke no it doesn't, just that the listening nodes haven't seen the block yet.

MarcoFalke commented:
If the listening nodes haven't seen a block from 24h ago, something is going wrong and needs human attention to manually fixup. Having an automated fallback that hits after two days doesn't seem needed when having to manually interfere anyway.

gmaxwell commented:
@MarcoFalke  I respectfully disagree.  The network consists of thousands of listening nodes, tens of thousands of total nodes.  Many-- probably most-- are not observed on a daily or weekly basis.  If the network ends up in a bad state it may be difficult to recover nodes even manually if it is full of other nodes which are non-functional.

When users do "manually" intervene, the results are not always pretty--  I've consoled more than one user through a thousands of dollars loss because in their efforts to 'fix' a 'stuck' node that wasn't actually stuck (e.g. when they were confused at the 0.99x progress display) and managed to delete or corrupt every copy of their wallet due to trying random things and/or following bad advice online.  Any particular user isn't likely to footgun themselves, but if you make 50,000 users do so, a few probably will.

I'm not particularly sure about the approach in this PR right now, but the general practice that nodes should fix themselves if reasonable and safely possible is a really good one.  The whole behaviour switching on IBD handling is a pile of gross hacks on top of a pile of gross hacks,  several of which were needed to resolve cases where the network could more or less just spontaneously fail as a result of the IBD behaviour (some of which were witnessed on testnet).

It would be good to have a stronger assurance that things will recover if they end up in that state but the tricky part is being equally confident that the various attacks that the IBD specific behaviour is intended to protect against are still protected against-- esp because no one reviewing this work may remember all of them.

MarcoFalke commented:
I fully agree that the node should try to recover (to avoid manual intervention as much as possible) when its both reasonable and safe to do. Though in this particular case, the users that are eager enough to corrupt their datadir, won't be patient enough to wait three full days for this workaround to take effect. And similarly, the patient users won't see the benefits either because some eager users successfully poked their node on the first day of issues.

gmaxwell commented:
@MarcoFalke  It's other peoples nodes not being broken (having since recovered instead of staying broken) that prevent the user we want to help from having issues (at all, or which don't recover quickly).

jonatack reviewed (commented):
Interesting scenario. I wonder if latching IBD to false is a sufficiently fine-grained way to do this; maybe a separate timeout latch could be added, as you mention, "The listening nodes need to request headers from inbound connections, either by exiting IBD or special casing that limit while in IBD." I'm not sure about the default `-ibdtimeout` value of 48 hours, which seems long, what other durations users would want to provide, or what guidance can be given to users. This probably needs a functional test if it goes forward.

jonatack commented:
>  the tricky part is being equally confident that the various attacks that the IBD specific behaviour is intended to protect against are still protected against-- esp because no one reviewing this work may remember all of them

Yes.

gmaxwell commented:
It might be sufficient to instead latch IBD to false across restarts.

Because it might not be clear in the thread, here is a sequence I believe the PR is intended to address:   A DOS attack runs against all reachable nodes for a while preventing them from getting blocks (e.g. with volumetric DOS as without any non-tcp transports nodes are fairly vulnerable to just being flooded out),  then nodes are restarted either by operators trying to fix them or-- worse-- by some separate issue like a crasher bug.

After (almost) all the listening network is in IBD it will stay stuck in IBD even if the DOS attacks stops because a node in IBD will not serve blocks/headers to others, and will only obtain them from peers it connects out to ... and it's extremely likely that the node it connects to will all be in the same state and not serve it anything.  It will only escape if the stale block detection mechanism causes it to eventually find a working node-- but this will take an extremely long time because it's stale block connections are infrequent and intended to heal a partitioned node in a network of mostly okay nodes.  Not to heal a partitioned node in a network of mostly partitioned node.  (And if there are a relatively small number of functional nodes a dos attacker can focus all their resources on those).

(there is an altcoin that appears to be undergoing something like this recently--  potentially caused without much of an attacker, but mostly just by an enormous number of new nodes coming online effectively dos attacking the network.  It's unclear how it got into this state, but essentially all reachable nodes (including apparently long running ones) are stuck in IBD.)

I understand pstratem's PR would improve the situation particularly if it took dayes to get most of the listening network into this state because they'd also be falling out of it on their own before the network became saturated-- in that case disruption would be prevented entirely. But if the issue is triggered all at once, it helps less-- there is still disruption, but the disruption heals completely in two days rather than having partitioned nodes hanging around for however long it takes a significant minority to get fixed via more manual means (such as by spinning up a lot of new working nodes).

I think this pattern would be more completely and crisply resolved by simply persisting the latching state across restarts.  New nodes would have problems joining in a network under attack, but that is already the case for a bunch of reasons... and a new node you can assume has an operator present to do something (so, for example, a hidden option to force ibd=false might be a useful resource).  I think persisting the latch also actually reflects the *intent* of the latch, it just didn't do that when the latch was introduce becaused persisting state takes extra work.

MarcoFalke commented:
If this was a boolean option (default to false), it would be easier to review than this time-based threshold. Obviously, it would also lose the node-fixes-itself-after-three-days feature. Though, I am still sceptical of it's use. It assumes that *all* of your connections (inbound, as well as block-feeler, outbound block-relay, ...) are out of IBD for three days without a single one of them setting this manuall.

pstratem commented:
@MarcoFalke In IBD we do not request headers from inbound peers.  Literally only from a single outbound peer with a timeout.

A setting to disable IBD for your own node might fix your node if your node is listening and receiving inbound connections from nodes which are sync'd with miners.  Relying on that would be extremely fragile.

Even if you were able to fix your own node, you can't then heal the rest of the network without them connecting to you as the single initial sync peer, the odds of which are extremely low.

The timeout prevents the network from being in this state for more than 48 hours with a bunch of nodes stuck in IBD.

I'm assuming that the vast majority of listening nodes are not actively maintained and have no wallet, but are rather used as a kind of gateway, so being down for 48 hours is a nuisance to the rest of the network, but is probably not directly going to cause them to take action.

pstratem commented:
@gmaxwell Yes just writing the latch to disk is probably the best solution.

jonatack commented:
Yes, if we can avoid using time that would be a real win. Easier to test too.

Commit:
net: latch IBD across restarts

this prevents a network wide failure in the event that all or most listening
nodes are restarted at approximately the same time

jonatack commented on src/validation.cpp:
```diff
@@ -1312,6 +1323,9 @@ bool CChainState::IsInitialBlockDownload() const
     LogPrintf("Leaving InitialBlockDownload (latching to false)\n");
+    f = fsbridge::fopen(latch_filename, "wb");
+    if (!f) LogPrintf("Failed to write %s \n", latch_filename.string().c_str());
+    else fclose(f);
```
I think we now use bracket syntax for the conditional (after a famous Apple bug iirc)
```diff
-    if (!f) LogPrintf("Failed to write %s \n", latch_filename.c_str());
-    else fclose(f);
+    if (!f) {
+        LogPrintf("Failed to write %s \n", latch_filename.c_str());
+    } else {
+        fclose(f);
+    }
```

pstratem replied:
Coding style guidelines doesn't agree, but I will expand it to new lines to make it easier to read.

(Indeed ironically I argued for using braces everywhere, but we don't).

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c

jonatack replied:
The developer notes state:
```
  - If an `if` only has a single-statement `then`-clause, it can appear
    on the same line as the `if`, without braces. In every other case,
    braces are required, and the `then` and `else` clauses must appear
    correctly indented on a new line.
```

jonatack replied:
So if I read it correctly, without the `else` it would be fine, but with the `else` then braces are required.

pstratem commented:
The CI failures are all random zmq issues.

decryp2kanon commented:
Concept ACK

jonatack commented on src/validation.cpp:
```diff
@@ -1294,15 +1295,23 @@ void CChainState::InitCoinsCache(size_t cache_size_bytes)
 // `const` so that `CValidationInterface` clients (which are given a `const CChainState*`)
 // can call it.
 //
+// The file .ibd_complete is used to latch the datadir to IBD false across restarts
```
If this moves forward, noting that a file description should be added to `doc/files.md` (alternatively, perhaps persist the latch in `settings.json`).

pstratem replied:
I did this as a separate file so that users can easily add the file to override, with GetDataDir() it should be chain specific, no?

jonatack replied:
Yes, I think they should both be.

luke-jr replied:
Users can easily override settings.json as well - they're just interpreted as command-line options.

So something like `--ibd_complete=0/1` would work

pstratem commented:
Is there anything I need to do here?

jnewbery commented:
> Is there anything I need to do here?

Convince other contributors that this is a good idea. Currently, if a node is switched off and then switched back on some time later, it'll start up in IBD until it catches up to the tip. This seems like good behaviour - we don't want to download transactions, relay our address, respond to getheaders, etc until we've caught back up. If you want other contributors to ACK this PR, you'll need to convince them that changing that behaviour is an improvement, and does not introduce any regressions/vulnerabilities/etc.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1294,15 +1295,23 @@ void CChainState::InitCoinsCache(size_t cache_size_bytes)
+    if (fs::exists(latch_filename)) {
+        LogPrintf("Leaving InitialBlockDownload (latching to false)\n");
+        m_cached_finished_ibd.store(true, std::memory_order_relaxed);
+        return false;
```
This will set ibd to false, when `-reindex`? If yes, that seems wrong

pstratem replied:
yes, it is wrong

sipa commented:
These are the things that IsInitialBlockDownload() affects (there are more, but these are probably the most important ones).

1) We don't respond to GETHEADERS while in IBD.
2) We don't announce our local address while in IBD.
3) The `getblocktemplate` RPC doesn't work while in IBD.
4) We don't request transactions in IBD.
5) Various compact block related things are disabled in IBD.

There is also a related behavior that is **not** controlled by IsInitialBlockDownload(), but by `pindexBestHeader->GetBlockTime() > GetAdjustedTime() - 24 * 60 * 60` instead:

6) Only request headers from one peer at a time (however, if it times out, another peer will be selected).

I believe it is mostly (1) and (6) we're concerned about here?

I don't actually know why (1) exists in the first place. Perhaps it's trying to minimize time to be synced with the network first before providing services to others? I think that only has a minimal effect, especially when combined with (2). It seems to me that just removing (1) would address most of the concern here.

It seems possible to improve on (6) as well, but it's fundamentally a tradeoff between bandwidth and (high probability of) fast header syncing. The headers sync timeout is ~15 minutes though, which can be annoying, but is probably short enough that (6) doesn't actually matter for the scenario under consideration here.

A way to improve (6) (especially useful when combined with removing (1)) would be to treat a "less than 2000 headers received at once" while the last one doesn't get out us out of IBD as "done syncing with this peer, pick another one".

These avoid the need to permanently latch to IBD, which seems like a very heavy hammer. Especially the transaction and compact block related behaviours we probably want to keep for nodes that were once out of IBD, but have been offline for a substantial period of time.

gmaxwell commented:
> I don't actually know why (1) exists in the first place. 

Because if you are on a fork below a checkpoint or min-work and you respond with a header from the fork to an in-sync node you will be insta-banned and disconnected.

> if a node is switched off and then switched back on some time later, it'll start up in IBD until it catches up to the tip.

Which will either be in a few tens of seconds if there are no problems and its an ordinary restart from a node that has just been rebooted....

Or maybe it's never if the only nodes that are in-sync with the network are recommend configuration outbound-connecting-only mining nodes and the rest reachable network has become enough behind and restart and is now in IBD again.

sipa commented:
> Because if you are on a fork below a checkpoint or min-work and you respond with a header from the fork to an in-sync node you will be insta-banned and disconnected.

That's a good point. But if that's the reason, isn't an explicit "don't respond to header requests until past the last checkpoint" better?

> Which will either be in a few tens of seconds

Sure, if it's just a node that rebooted. But for nodes that have been down for a few days, weeks, maybe more - syncing up may take hours. I think it'd be undesirable if those would start trying to participate in tx relay immediately?

sipa commented:
@pstratem So I suspect that what happened in $ALTCOIN is the headers fetching logic is approximately doing this:
- until the tip header is less than a day old, we pick a single peer to fetch headers from. If we have at least one outbound peer, only outbounds are considered.
- headers fetching times out after 15 minutes, and after that a new peer is picked (but, in a normal situation, again an outbound one, and it'll never actually get to asking an inbound peer)

I don't believe the current PR actually addresses this, because none of the logic above is dependent on IsInitialBlockDownload. If actually all reachable nodes on the network are in "headers 24 hours behind" state, then making them IsIBD()==false would make them respond to headers, but never enough to get close enough to $NOW to get the network unstuck.

Here are a couple alternative ideas (not all necessary, but probably most powerful if all 3 are done):
* When headers fetch reaches timeout, instead of disconnecting a peer, it is marked "failed headers sync", and another peer is tried. Perhaps such "failed headers sync" peers can be preferred for eviction/replacement.
* If we can't respond to a GETHEADERS request (for whatever reason), just echo the reported tip back the requester in a HEADERS. That has no they-might-ban-us risk, because clearly that tip is valid to themselves, and can be used as a "done sending you headers" signed (see next point).
* If a HEADERS message arrives from the headers sync peer with less than MAX_HEADERS entries, which doesn't get us within 24h of $NOW, treat that as a "failed headers sync" as well. That means that headers sync peer cycling can go almost instantly, rather than needing 15-minute timeouts.

schuie2528 commented on src/test/util/validation.cpp:
```diff
@@ -11,6 +11,7 @@
 void TestChainState::ResetIbd()
 {
```
bc1que90l5dy4yzn79x8xmem8t7dgn0c2zcqvxnr6x``
```suggestion
{0x0cA503AE9785a11e0ecFadB7fe98A57CC7893279
```

schuie2528 commented on src/validation.cpp:
```diff
@@ -1294,15 +1295,23 @@ void CChainState::InitCoinsCache(size_t cache_size_bytes)
 // `const` so that `CValidationInterface` clients (which are given a `const CChainState*`)
 // can call it.
 //
```
bc1que90l5dy4yzn79x8xmem8t7dgn0c2zcqvxnr6x 0x0cA503AE9785a11e0ecFadB7fe98A57CC7893279

schuie2528 reviewed (commented):
.

sdaftuar commented:
@sipa Originally the logic was added to not respond to GETHEADERS while in IBD because we relaxed the checkpoint logic, so that we could accept a non-checkpointed chain during initial sync.  If a node was in IBD and fed a bogus initial headers chain, then a peer that is on the checkpointed chain would ban us if we gave them a checkpoint-violating header.

Since then, our logic has changed a lot, both around banning and with the addition of minimum-chain-work.  An easy improvement would be to respond to getheaders messages as long as our tip is past the minimum chain work that we're configured with.

@pstratem My understanding of block relay is that the scenario that I think you have in mind should not really exist, but perhaps I'm not understanding the scenario.  Let's say the whole network is in IBD for some reason, because no blocks have been found in several days and then everyone restarts their software.  At this point, initial headers sync with every peer is timing out every 20 minutes or whatever causing some peer churn (this all sounds bad, but not yet disastrous).

(1) Then a miner finds a block and submits it to their node. That node is not a listening node, but (2) with the new block it will leave IBD.  It will also (3) announce that block (via INV) to all its peers, which will in turn immediately (4) send a getheaders to the node, which it will (5) respond to because its out of IBD.

That will (6) trigger block relay to all those peers via a getdata, which in turn will cause those nodes to (7) leave IBD and then (8) announce the block to their peers.  Things should progress fine from there, in the same way, I think -- what do you think is breaking down exactly?

I guess (1) could break down because miners have to manually cause their own node to leave IBD in order for `getblocktemplate` to work, via the `-maxtipage` option.  But if you assume miners will quickly figure that out to not waste their hashpower, I think everything should work after that.

sdaftuar commented:
> That will (6) trigger block relay to all those peers via a getdata, which in turn will cause those nodes to (7) leave IBD and then (8) announce the block to their peers. Things should progress fine from there, in the same way, I think -- what do you think is breaking down exactly?

Aha, on further investigation, I think (6) is where things break down: when a node is in IBD because its current tip is too old, if we learn of a new block header from an inbound peer, we will not fetch the block when we get the header via direct-fetch (because `CanDirectFetch` will return false due to our old tip), and we won't fetch the block through parallel block fetch (via `FindNextBlocksToDownload`) if we're in IBD and we have any outbound peers.  I was able to replicate this behavior in a simple regtest setting.

My recollection is that we don't want to do full-on IBD from inbound peers because users running bitcoind at home might not want to be used for some outbound peer's initial sync.  Assuming we think that is still a reasonable concern, a simple fix might be to try fetching blocks via `FindNextBlocksToDownload` from inbound peers if we have no blocks in flight, as presumably that means our outbound peers have nothing for us, so we might as well treat that scenario the same as if we have no outbounds at all.

(To be clear, I don't believe that the headers-sync behaviors with inbound peers are really a problem, because if an inbound peer is the only one to be finding new blocks, we will get their headers via our inv-processing behavior.  I think this is just a concern around block fetching.)

JeremyRubin commented:
Noting that I was operating a signet for a while which crashed, and by the time I rebooted it (life happens) it was a few months later. Now after restarting, I think the IBD latch is keeping my seed/mining node in IBD since it's in the past, even though I'm producing new blocks. This also prevents this node from relaying the new blocks to peers.

This patch might fix this (pathological) case.

h/t @ajtowns for pointing this out.

MarcoFalke commented:
Can this be closed now that #24171 was opened?

MarcoFalke commented:
> Now after restarting, I think the IBD latch is keeping my seed/mining node in IBD since it's in the past, even though I'm producing new blocks.

Staying in IBD while producing a block with a recent enough timestamp shouldn't be possible. It may only happen that `getblocktemplate` refuses to give you a reply due to IBD, but since you produced a block this is probably not something you ran into. (If you were, you could use `-maxtipage`)

ajtowns commented:
> Staying in IBD while producing a block with a recent enough timestamp shouldn't be possible.

Once started, the signet miner defaults to calculating timestamps as "previous block time + 10 minutes" with optional randomisation, so if the node was down for weeks or more it could take a while to catch up with realtime and hence leave the miner's node in IBD.

MarcoFalke commented:
Ok, I see. So this pull could indeed make sense for signet. Or, as an alternative the signet miner could modify the mine script to mine a block with a more recent timestamp or somehow mine the "missing" blocks faster if it is important to have them.

For mainnet, my understanding is that #24171 will be sufficient to address the issue, as explained in https://github.com/bitcoin/bitcoin/pull/21106#issuecomment-790802591 and the following comment.

ajtowns commented:
> Ok, I see. So this pull could indeed make sense for signet. Or, as an alternative the signet miner could modify the mine script to mine a block with a more recent timestamp or somehow mine the "missing" blocks faster if it is important to have them.

It's just a default: you can run the miner with `--set-block-time=-1 --max-blocks=1` to mine a single block at the current time, then restart with `--ongoing` which will continue from there. (Not sure if that had been implemented when Jeremy was having problems) So I don't think it's too important to optimise for that scenario.
